src/aws_encryption_sdk/keyring/multi_keyring.py

```diff
@@ -0,0 +1,107 @@
+
+    children = attr.ib(
+        default=attr.Factory(tuple), validator=optional(deep_iterable(member_validator=instance_of(Keyring)))
+    )
+    generator = attr.ib(default=None, validator=optional(instance_of(Keyring)))
+
+    def __attrs_post_init__(self):
+        # type: () -> None
+        """Prepares initial values not handled by attrs."""
+        neither_generator_nor_children = self.generator is None and not self.children
+        if neither_generator_nor_children:
+            raise TypeError("At least one of generator or children must be provided")
+
+        _generator = (self.generator,) if self.generator is not None else ()
+        self._decryption_keyrings = itertools.chain(_generator, self.children)
+
+    def on_encrypt(self, encryption_materials):
+        # type: (EncryptionMaterials) -> EncryptionMaterials
+        """Generate a data key using generator keyring
+        and encrypt it using any available wrapping key in any child keyring.
+
+        :param encryption_materials: Encryption materials for keyring to modify.
+        :type encryption_materials: aws_encryption_sdk.materials_managers.EncryptionMaterials
+        :returns: Optionally modified encryption materials.
+        :rtype: aws_encryption_sdk.materials_managers.EncryptionMaterials
+        :raises EncryptKeyError: if unable to encrypt data key.
+        """
+        # Check if generator keyring is not provided and data key is not generated
+        if self.generator is None and not encryption_materials.data_encryption_key:
+            raise EncryptKeyError(
+                "Generator keyring not provided "
+                "and encryption materials do not already contain a plaintext data key."
+            )
+
+        # Call on_encrypt on the generator keyring if it is provided
+        if self.generator is not None:
+            encryption_materials = self.generator.on_encrypt(encryption_materials)
+
+        # Check if data key is generated
+        if not encryption_materials.data_encryption_key:
+            raise GenerateKeyError("Unable to generate data encryption key.")
+
+        # Call on_encrypt on all other keyrings
+        for keyring in self.children:
+            encryption_materials = keyring.on_encrypt(encryption_materials)
+
+        return encryption_materials
+
+    def on_decrypt(self, decryption_materials, encrypted_data_keys):
+        # type: (DecryptionMaterials, Iterable[EncryptedDataKey]) -> DecryptionMaterials
+        """Attempt to decrypt the encrypted data keys.
+
+        :param decryption_materials: Decryption materials for keyring to modify.
+        :type decryption_materials: aws_encryption_sdk.materials_managers.DecryptionMaterials
+        :param encrypted_data_keys: List of encrypted data keys.
+        :type: List of `aws_encryption_sdk.structures.EncryptedDataKey`
+        :returns: Optionally modified decryption materials.
+        :rtype: aws_encryption_sdk.materials_managers.DecryptionMaterials
+        """
```

Comment:
I've been trying to think of a way that we could simplify this function. It occurred to me today as I was reading through this that if we swap the order of operations in the loop, we can get rid of the initial check.

ie:
```python
        for keyring in self._decryption_keyrings:
            if decryption_materials.data_encryption_key:
                return decryption_materials
            decryption_materials = keyring.on_decrypt(decryption_materials, encrypted_data_keys)

        return decryption_materials
```

This would still handle each case correctly, and would be slightly fewer lines of code to maintain.
However, I'm not sure whether or not this would be an overly clever premature optimization.
I'm not opposed to keeping the extra initial check if it aids the readability.

@MeghaShetty I'm interested in your perspective on this. Which do you think is more readable?

Reply:
This code is readable as well. It is easy to understand. I like this modification! I'll make the changes.